### ISSUE TITLE
Expand and enforce linting guidelines

### ISSUE THREAD

**Issue by Amorymeltzer** [azatoth/twinkle#500]
Let's talk about linting.  As noted in the docs, there isn't really a consistent style applied to Twinkle.  That's okay!  But with around 17,000 lines of code, maybe it's also okay to have some more options in place.  As of #400 we have the basics via `eslint:recommended`, and `no-caller` was added in cd45282 (`max-len` is listed but unused).

I went through the [eslint rules](https://eslint.org/docs/rules/) and picked out a bunch of possibilities; the list is a combination of options that we already follow, options we don't, options I just personally prefer, and options that I thought might seem reasonable to at least someone out there.  Some are listed with different configurations.  This is a _long_ list; still, it is by no means meant to be exhaustive or limiting, so feel free to proffer others.  Many are probably unnecessary or overkill.

I generated a table with the current results from each, arranged alphabetically by category, and intend to update the ✔️ or ❌ marks as this conversation goes forward.  The links will take you to the online docs page for each option, which should explain each in full.  I've made some (heavy-handed?) suggestions for a lot of them, in particular those with multiple configurations listed, to get us started, but feel free to disagree with those choices.

This isn't particularly urgent (see also #477), and there are a number of PRs we should review/merge beforehand to avoid conflicts, but once we've got a firm list I'll go through and make the changes.

cc @MusikAnimal @atlight @kevinji @azatoth

| Use?      | Option      | Config      | Curr errors | Fix?    | Category    | Notes?      |
| ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- |
| ❌  | [no-extra-parens](https://eslint.org/docs/rules/no-extra-parens) | "error" | 146 | Yes | Possible errors | Alt|
| ❌  |   [no-extra-parens](https://eslint.org/docs/rules/no-extra-parens) | ["error", "functions"] | 0 | Yes | Possible errors | Alt; might be allow additional clarity in ternaries|
| ✔️  |   [no-extra-parens](https://eslint.org/docs/rules/no-extra-parens) | ["error", "all", { "nestedBinaryExpressions": false }] | 113 | Yes | Possible errors | Alt; required for `no-mixed-operators` |
| ✔️  |   [no-prototype-builtins](https://eslint.org/docs/rules/no-prototype-builtins) | "error" | 1 | No |  Possible errors| |
| ✔️  |   [block-scoped-var](https://eslint.org/docs/rules/block-scoped-var) | "error" | 0 | No | Best practices | |
| ✔️  |   [curly](https://eslint.org/docs/rules/curly) | "error" | 20 | Yes | Best practices | Alt|
| ❌  |   [curly](https://eslint.org/docs/rules/curly) | ["error", "multi"] | 787 | Yes | Best practices | Alt|
| ❌  |   [curly](https://eslint.org/docs/rules/curly) | ["error", "multi-line"] | 5 | Yes | Best practices | Alt|
| ✔️  |   [default-case](https://eslint.org/docs/rules/default-case) | "error" | 2 | No | Best practices | |
| ✔️  |   [eqeqeq](https://eslint.org/docs/rules/eqeqeq) | "error" | 8 | Mixed | Best practices | Alt|
| ❌  |   [eqeqeq](https://eslint.org/docs/rules/eqeqeq) | ["error", "smart"] | 8 | Mixed | Best practices | Alt|
| ✔️  |   [guard-for-in](https://eslint.org/docs/rules/guard-for-in) | "error" | 0 | No | Best practices | |
| ✔️  |   [no-else-return](https://eslint.org/docs/rules/no-else-return) | "error" | 6 | Yes | Best practices | |
| ❌  |   [no-empty-function](https://eslint.org/docs/rules/no-empty-function) | "error" | 2 | No | Best practices | |
| ❌  |   [no-eq-null](https://eslint.org/docs/rules/no-eq-null) | "error" | 0 | No |  Best practices| |
| ❔  |   [no-global-assign](https://eslint.org/docs/rules/no-global-assign) | "error" | 0 | No | Best practices | See also `no-undefined` and `no-shadow-restricted-names`|
| ✔️  |   [no-implicit-coercion](https://eslint.org/docs/rules/no-implicit-coercion) | ["error", { "boolean": false }] | 0 | Yes | Best practices | Alt|
| ❌  |   [no-implicit-coercion](https://eslint.org/docs/rules/no-implicit-coercion) | ["error"] | 18 | Yes | Best practices | Alt|
| ✔️  |   [no-lone-blocks](https://eslint.org/docs/rules/no-lone-blocks) | "error" | 0 | No | Best practices | |
| ✔️  |   [no-multi-spaces](https://eslint.org/docs/rules/no-multi-spaces) | ["error", { "ignoreEOLComments": true }] | 17 | Yes | Best practices | Alt|
| ❌  |   [no-multi-spaces](https://eslint.org/docs/rules/no-multi-spaces) | ["error"] | 251 | Yes | Best practices | Alt|
| ❔  |   [no-script-url](https://eslint.org/docs/rules/no-script-url) | "error" | 3 | No | Best practices | |
| ✔️  |   [no-useless-catch](https://eslint.org/docs/rules/no-useless-catch) | "error" | 0 | No | Best practices | |
| ✔️  |   [no-useless-return](https://eslint.org/docs/rules/no-useless-return) | "error" | 0 | Yes | Best practices | |
| ✔️  |   [yoda](https://eslint.org/docs/rules/yoda) | "error" | 0 | Yes | Best practices | |
| ❔  |   [no-shadow-restricted-names](https://eslint.org/docs/rules/no-shadow-restricted-names) | "error" | 2 | No | Variables | Dislikes the format of the anonymous functions in `twinkle.js` and `morebits.js`; see also `no-undefined`|
| ❔  |   [no-undefined](https://eslint.org/docs/rules/no-undefined) | "error" | 24 | No | Variables | Alternatively, can use `no-global-assign` and `no-shadow-restricted-names` to allowed usage of `undefined` while still being safe, but either way, they dislike the format of the anonymous functions in `twinkle.js` and `morebits.js`|
| ❔  |   [array-bracket-spacing](https://eslint.org/docs/rules/array-bracket-spacing) | ["error", "never"] | 127 | Yes | Stylistic | |
| ❔  |   [array-bracket-spacing](https://eslint.org/docs/rules/array-bracket-spacing) | ["error", "never", { "objectsInArrays": true }] | 126 | Yes | Stylistic | |
| ❔  |   [array-bracket-spacing](https://eslint.org/docs/rules/array-bracket-spacing) | ["error", "always"] | 52 | Yes | Stylistic | |
| ✔️  |   [block-spacing](https://eslint.org/docs/rules/block-spacing) | "error" | 0 | Yes | Stylistic | |
| ❌  |   [brace-style](https://eslint.org/docs/rules/brace-style) | ["error", "1tbs", { "allowSingleLine": true }] | 33 | Yes (27) | Stylistics | Alt|
| ✔️  |   [brace-style](https://eslint.org/docs/rules/brace-style) | ["error", "1tbs"] | 75 | Yes (69) | Stylistic | Alt|
| ✔️  |   [comma-dangle](https://eslint.org/docs/rules/comma-dangle) | "error" | 5 | Yes | Stylistic | |
| ❌  |   [comma-dangle](https://eslint.org/docs/rules/comma-dangle) | ["error", "always-multiline"] | 1326 | Yes | Stylistic | |
| ✔️  |   [comma-spacing](https://eslint.org/docs/rules/comma-spacing) | ["error", { "before": false, "after": true }] | 22 | Yes | Stylistic | |
| ✔️  |   [computed-property-spacing](https://eslint.org/docs/rules/computed-property-spacing) | ["error", "never"] | 42 | Yes | Stylistic | |
| ✔️  |   [func-call-spacing](https://eslint.org/docs/rules/func-call-spacing) | ["error", "never"] | 4 | Yes | Stylistic | |
| ❌  |   [indent](https://eslint.org/docs/rules/indent) | ["error", "tab"] | 14604 | Yes | Stylistic | Alt; mostly due to not indenting after the wrapper function in each module|
| ❌  |   [indent](https://eslint.org/docs/rules/indent) | ["error", "tab", { "outerIIFEBody": 0 }] | 4484 | Yes | Stylistic | Alt; allow 0 indent after wrapper function, seems reasonable|
| ✔️  |   [indent](https://eslint.org/docs/rules/indent) | ["error", "tab", { "outerIIFEBody": 0, "SwitchCase": 1 }] | 3303 | Yes | Stylistic | Alt; enforce a tab before `case` statements|
| ❌  |   [indent](https://eslint.org/docs/rules/indent) | ["error", "tab", { "outerIIFEBody": 0, "ArrayExpression": "off", "ObjectExpression": "off" }] | 2082 | Yes | Stylistic | Alt; allow Twinkle to use 2 indents instead of 1 in multi-line arrays and objects|
| ❌  |   [indent](https://eslint.org/docs/rules/indent) | ["error", "tab", { "outerIIFEBody": 0, "SwitchCase": 1, "ArrayExpression": "off", "ObjectExpression": "off" }] | 1201 | Yes | Stylistic | Alt; all of the above|
| ❌  |   [key-spacing](https://eslint.org/docs/rules/key-spacing) | ["error", {"singleLine": {"beforeColon": true, "afterColon": true}}] | 1420 | Yes | Stylistic | Alt|
| ❌  |   [key-spacing](https://eslint.org/docs/rules/key-spacing) | ["error", {"singleLine": {"beforeColon": true, "afterColon": false}}] | 2476 | Yes | Stylistic | Alt|
| ✔️  |   [key-spacing](https://eslint.org/docs/rules/key-spacing) | ["error", {"singleLine": {"beforeColon": false, "afterColon": true}}] | 212 | Yes | Stylistic | Alt|
| ❌  |   [key-spacing](https://eslint.org/docs/rules/key-spacing) | ["error", {"singleLine": {"beforeColon": false, "afterColon": false}}] | 1268 | Yes | Stylistic | Alt|
| ✔️  |   [keyword-spacing](https://eslint.org/docs/rules/keyword-spacing) | ["error", { "after": true , "before": true}] | 454 | Yes | Stylistic | Alt|
| ❌  |   [keyword-spacing](https://eslint.org/docs/rules/keyword-spacing) | ["error", { "after": true , "before": false}] | 734 | Yes | Stylistic | Alt|
| ❌  |   [keyword-spacing](https://eslint.org/docs/rules/keyword-spacing) | ["error", { "after": false , "before": true}] | 1333 | Yes | Stylistic | Alt|
| ❌  |   [keyword-spacing](https://eslint.org/docs/rules/keyword-spacing) | ["error", { "after": false , "before": false}] | 1613 | Yes | Stylistic | Alt|
| ✔️  |   [linebreak-style](https://eslint.org/docs/rules/linebreak-style) | ["error", "unix"] | 0 | Yes | Stylistic | |
| ❌  |   [multiline-comment-style](https://eslint.org/docs/rules/multiline-comment-style) | ["error", "separate-lines"] | 24 | Yes | Stylistic |Alt; probably fine to mix-and-match these |
| ❌  |   [multiline-comment-style](https://eslint.org/docs/rules/multiline-comment-style) | ["error", "bare-block"] | 130 | Yes | Stylistic | Alt; probably fine to mix-and-match these|
| ❌  |   [multiline-comment-style](https://eslint.org/docs/rules/multiline-comment-style) | ["error", "starred-block"] | 127 | Yes | Stylistic | Alt; probably fine to mix-and-match these|
| ✔️  |   [no-array-constructor](https://eslint.org/docs/rules/no-array-constructor) | "error" | 0 | No | Stylistic | |
| ✔️  |   [no-bitwise](https://eslint.org/docs/rules/no-bitwise) | "error" | 0 | No |  Stylistic| |
| ❌  |   [no-lonely-if](https://eslint.org/docs/rules/no-lonely-if) | "error" | 6 | Yes (5) | Stylistic | |
| ✔️  |   [no-mixed-operators](https://eslint.org/docs/rules/no-mixed-operators) | "error" | 6 | No | Stylistic | |
| ❔  |   [no-multiple-empty-lines](https://eslint.org/docs/rules/no-multiple-empty-lines) | "error" | 24 | Yes | Stylistic | Default is 2; all but two instances are in morebits, the other are in twinkle.js and xfd |
| ❔  |   [no-multiple-empty-lines](https://eslint.org/docs/rules/no-multiple-empty-lines) | ["error", {"max": 3}] | 0 | Yes | Stylistic | Might be nice for setting apart functions in morebits |
| ✔️  |   [no-new-object](https://eslint.org/docs/rules/no-new-object) | "error" | 0 | No | Stylistic |  | 
| ❌  |   [no-tabs](https://eslint.org/docs/rules/no-tabs) | ["error"] | 17653 | No | Stylistic | Alt; lolno|
| ✔️  |   [no-tabs](https://eslint.org/docs/rules/no-tabs) | ["error", { "allowIndentationTabs": true }] | 11 | No | Stylistic |Alt; will complain about indentation tabs in commented code, though (but we should just remove those) |
| ✔️  |   [no-trailing-spaces](https://eslint.org/docs/rules/no-trailing-spaces) | "error" | 2 | Yes | Stylistic | |
| ✔️  |   [no-unneeded-ternary](https://eslint.org/docs/rules/no-unneeded-ternary) | "error" | 0 | Yes | Stylistic | |
| ✔️  |   [no-whitespace-before-property](https://eslint.org/docs/rules/no-whitespace-before-property) | "error" | 0 | Yes | Stylistic | |
| ❔  |   [operator-linebreak](https://eslint.org/docs/rules/operator-linebreak) | ["error", "after"] | 0 | Yes | Stylistic | |
| ❌  |   [quote-props](https://eslint.org/docs/rules/quote-props) | ["error", "always"] | 5219 | Yes | Stylistic | |
| ❔  |   [quote-props](https://eslint.org/docs/rules/quote-props) | ["error", "as-needed"] | 500 | Yes | Stylistic | |
| ❔  |   [quote-props](https://eslint.org/docs/rules/quote-props) | ["error", "consistent"] | 2 | Yes | Stylistic | |
| ❔  |   [quote-props](https://eslint.org/docs/rules/quote-props) | ["error", "consistent-as-needed"] | 393 | Yes | Stylistic | |
| ❔  |   [quote-props](https://eslint.org/docs/rules/quote-props) | ["error", "consistent-as-needed", {"keywords": true}] | 387 | Yes | Stylistic |  |
| ✔️  |   [quotes](https://eslint.org/docs/rules/quotes) | ["error", "single", { "avoidEscape": true }] | 4532 | Yes | Stylistic | Alt; Slight preference (56-44)|
| ❌  |   [quotes](https://eslint.org/docs/rules/quotes) | ["error", "double", { "avoidEscape": true }] | 5790 | Yes | Stylistic | Alt; slight underdog (44-56)|
| ✔️  |   [semi](https://eslint.org/docs/rules/semi) | ["error", "always"] | 2 | Yes | Stylistic | |
| ✔️  |   [space-before-blocks](https://eslint.org/docs/rules/space-before-blocks) | "error" | 60 | Yes | Stylistic | Complains about the current style of the wrapper function for each module|
| ✔️  |   [space-in-parens](https://eslint.org/docs/rules/space-in-parens) | ["error", "never"] | 4452 | Yes | Stylistic | Alt; about time we committed to one style|
| ❌  |   [space-in-parens](https://eslint.org/docs/rules/space-in-parens) | ["error", "always"] | 9340 | Yes | Stylistic | Alt|
| ✔️  |   [space-infix-ops](https://eslint.org/docs/rules/space-infix-ops) | "error" | 67 | Yes | Stylistic | |
| ✔️  |   [space-unary-ops](https://eslint.org/docs/rules/space-unary-ops) | "error" | 20 | Yes | Stylistic | |
| ❌  |   [spaced-comment](https://eslint.org/docs/rules/spaced-comment) | ["error", "never"] | 1168 | Yes | Stylistic | Alt|
| ✔️  |   [spaced-comment](https://eslint.org/docs/rules/spaced-comment) | ["error", "always", { "line": { "exceptions": ["-"] }, "block": { "balanced": true } }] | 86 | Yes | Stylistic | Alt; would require a space before \<nowiki\> tags|
| ❌  |   [spaced-comment](https://eslint.org/docs/rules/spaced-comment) | ["error", "always", { "line": { "exceptions": ["-", "\<nowiki\>"] }, "block": { "balanced": true } }] | 64 | Yes | Stylistic | Alt; don't require a space before \<nowiki\> tags|
| ✔️  |   [switch-colon-spacing](https://eslint.org/docs/rules/switch-colon-spacing) | "error" | 0 | Yes | Stylistic | |

<details>
  <summary>Updated .eslintrc.json with the above options</summary>
  
```json
{
	"extends": ["eslint:recommended"],
	"root": true,
	"env": {
		"browser": true,
		"jquery": true
	},
	"globals": {
		"mw": true,
		"moment": true
	},
	"rules": {
		"no-extra-parens": ["error", "all", { "nestedBinaryExpressions": false }],
		"no-prototype-builtins": "error",
		"block-scoped-var": "error",
		"curly": "error",
		"default-case": "error",
		"eqeqeq": "error",
		"guard-for-in": "error",
		"no-caller": "error",
		"no-else-return": "error",
		"no-implicit-coercion": ["error", { "boolean": false }],
		"no-lone-blocks": "error",
		"no-multi-spaces": ["error", { "ignoreEOLComments": true }],
		"no-useless-catch": "error",
		"no-useless-return": "error",
		"yoda": "error",
		"block-spacing": "error",
		"brace-style": ["error", "1tbs"],
		"comma-dangle": "error",
		"comma-spacing": ["error", { "before": false, "after": true }],
		"computed-property-spacing": ["error", "never"],
		"func-call-spacing": ["error", "never"],
		"indent": ["error", "tab", { "outerIIFEBody": 0, "SwitchCase": 1 }],
		"key-spacing": ["error", {"singleLine": {"beforeColon": false, "afterColon": true}}],
		"keyword-spacing": ["error", { "after": true , "before": true}],
		"linebreak-style": ["error", "unix"],
		"no-array-constructor": "error",
		"no-bitwise": "error",
		"no-mixed-operators": "error",
		"no-new-object": "error",
		"no-tabs": ["error", { "allowIndentationTabs": true }],
		"no-trailing-spaces": "error",
		"no-unneeded-ternary": "error",
		"no-whitespace-before-property": "error",
		"quotes": ["error", "single", { "avoidEscape": true }],
		"semi": ["error", "always"],
		"space-before-blocks": "error",
		"space-in-parens": ["error", "never"],
		"space-infix-ops": "error",
		"space-unary-ops": "error",
		"spaced-comment": ["error", "always", { "line": { "exceptions": ["-"] }, "block": { "balanced": true } }],
		"switch-colon-spacing": "error"
	}
}

```
</details>

**Comment by kevinji** [comment]
I think that all of your suggestions sound fine; I would also probably enable most of the ? rules you had above with < 10 issues and just fix those issues too.

I have a slight preference for these options:

- [curly](https://eslint.org/docs/rules/curly): "all" (default) -- seems easy to fix
- [comma-dangle](https://eslint.org/docs/rules/comma-dangle): "always-multiline" -- this looks a bit nicer and all browsers these days should support this
- [eqeqeq](https://eslint.org/docs/rules/eqeqeq): "always" (default) -- seems easy enough to fix the current issues
- [indent](https://eslint.org/docs/rules/indent): the chosen `["error", "tab", { "outerIIFEBody": 0, "SwitchCase": 1 }]` seems fine
- [no-implicit-coercion](https://eslint.org/docs/rules/no-implicit-coercion): default -- it seems like we can probably remove the few cases of `!!`
- [no-lonely-if](https://eslint.org/docs/rules/no-lonely-if): "error" (default) -- seems easy to fix
- [space-before-blocks](https://eslint.org/docs/rules/space-before-blocks): "always" (default) -- seems easy to fix
- [space-in-parens](https://eslint.org/docs/rules/space-in-parens): chosen `["error", "never"]` -- seems more widely used than "always" in most style guides now

**Comment by Amorymeltzer** [comment]
@kevinji Thanks for chiming in here.  I left a lot of them as ❔ so as not to unilaterally impose my across the board, but as time has gone on I've slowly opined on more.  I also figure we can be more exacting on the items that can be automatically fixed.  To reply to your points that differed from the WIP above:

- [eqeqeq](https://eslint.org/docs/rules/eqeqeq): Yeah, the default is better.  I was thinking we could avoid unnecessary strictness, but you're probably right that it's best to be consistent.  Plus it means we can skip `no-eq-null`!  ✔️Done.
- [no-implicit-coercion](https://eslint.org/docs/rules/no-implicit-coercion): The only one we regularly use is `!!`, and I rather dislike `Boolean`, but I'll get over it and it's automatically fixable so there should be minimum impact anyway.  ✔️Done.
- [space-before-blocks](https://eslint.org/docs/rules/space-before-blocks): Agreed; not sure why I had the qualifier there.  ✔️Done.
- [comma-dangle](https://eslint.org/docs/rules/comma-dangle): "always-multiline" -- I'm less sold on this one.  It's a pretty clear change from what we nearly universally do (1326 for always-multiline versus 5 for the default).  It certainly helps with diffs, which is probably enough of a reason, but this is one area where we very nearly do have a house style already.
- [no-lonely-if](https://eslint.org/docs/rules/no-lonely-if): My only hesitation here is that it makes things like the following:
````javascript
if (cond) {
    if (cond) {
        ...
    } else {
        ...
    }
else {
    if (cond) {
        ...
    } else {
        ...
    }
}
````
Into:
````javascript
if (cond) {
    if (cond) {
        ...
    } else {
        ...
    }
else if (cond) {
    ...
} else {
    ...
}
````
In the abstract it's fine, but it can ruin parallelism and be generally less clear.  Not a major issue and not something I'm against, but that's my hesitation.

**Comment by kevinji** [comment]
- [comma-dangle](https://eslint.org/docs/rules/comma-dangle): Ah, I didn't realize what the stats were. I'm happy to keep it as the setting you chose then.
- [no-lonely-if](https://eslint.org/docs/rules/no-lonely-if): Your comment sounds reasonable; we can probably keep this rule off for now.

**Comment by Amorymeltzer** [comment]
Lacking further input (and wanting to do get this done before working on #477, which is more important) I'm going to move forward (slowly) with we've got here.  I don't think we'll be as popular as airbnb's setup, but this should work for us.  There are a few disparate items on this list I'd like to get done, but it's certainly not urgent.

Barring any objections (@azatoth @MusikAnimal @atlight @kevinji and @siddharthvp and @DannyS712) I've opened #611 to track the PRs I want to get reviewed and merged before going ahead, hoping to avoid conflicts.
